File: uber/uber_algo/LC113PathSumII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        if not root:
            return []
        answer = []
        self.dfs(answer, root, sum, [])
        return answer

    def dfs(self, answer, node, target, path):
        if not node:
            return
        if node and not node.left and not node.right:
            path.append(node.val)
            if sum(path) == target:
                answer.append(path)
            return
        self.dfs(answer, node.left, target, path+[node.val])
        self.dfs(answer, node.right, target, path+[node.val])

File: apple/LC113PathSumII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> answer = new ArrayList<>();
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        if (root == null) return answer;
        Stack<Integer> path = new Stack<>();
        dfs(root, path, sum);
        return answer;
    }

    public void dfs(TreeNode root, Stack<Integer> path, int sum) {
        path.push(root.val);
        if (root.left == null && root.right == null) {
            if (sum == root.val) {
                answer.add(new ArrayList<>(path));
            }
        }
        if (root.left != null) dfs(root.left, path, sum-root.val);
        if (root.right != null) dfs(root.right, path, sum-root.val);
        path.pop();
    }
}
